py2puml/parsing/astvisitors.py

```diff
@@ -0,0 +1,208 @@
+    Attribute, Name, Tuple as AstTuple, Subscript, get_source_segment
+)
+from collections import namedtuple
+
+from py2puml.domain.umlclass import UmlAttribute
+from py2puml.domain.umlrelation import UmlRelation, RelType
+from py2puml.parsing.compoundtypesplitter import CompoundTypeSplitter, SPLITTING_CHARACTERS
+from py2puml.parsing.moduleresolver import ModuleResolver, NamespacedType
+
+Variable = namedtuple('Variable', ['id', 'type_expr'])
+
+class SignatureVariablesCollector(NodeVisitor):
+    '''
+    Collects the variables and their type annotations from the signature of a constructor method
+    '''
+    def __init__(self, constructor_source: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.constructor_source = constructor_source
+        self.class_self_id: str = None
+        self.variables: List[Variable] = []
+
+    def visit_arg(self, node: arg):
+        variable = Variable(node.arg, node.annotation)
+
+        # first constructor variable is the name for the 'self' reference
+        if self.class_self_id is None:
+            self.class_self_id = variable.id
+        # other arguments are constructor parameters
+        else:
+            self.variables.append(variable)
+
+
+class AssignedVariablesCollector(NodeVisitor):
+    '''Parses the target of an assignment statement to detect whether the value is assigned to a variable or an instance attribute'''
+    def __init__(self, class_self_id: str, annotation: expr):
+        self.class_self_id: str = class_self_id
+        self.annotation: expr = annotation
+        self.variables: List[Variable] = []
+        self.self_attributes: List[Variable] = []
+
+    def visit_Name(self, node: Name):
+        '''
+        Detects declarations of new variables
+        '''
+        if node.id != self.class_self_id:
+            self.variables.append(Variable(node.id, self.annotation))
+
+    def visit_Attribute(self, node: Attribute):
+        '''
+        Detects declarations of new attributes on 'self'
+        '''
+        if isinstance(node.value, Name) and node.value.id == self.class_self_id:
+            self.self_attributes.append(Variable(node.attr, self.annotation))
+
+    def visit_Subscript(self, node: Subscript):
+        '''
+        Assigns a value to a subscript of an existing variable: must be skipped
+        '''
+        pass
+
+
+class ConstructorVisitor(NodeVisitor):
+    '''Identifies the attributes (and infer their type) assigned to self in the body of a constructor method'''
+    def __init__(self, constructor_source: str, class_name: str, root_fqdn: str, module_resolver: ModuleResolver, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.constructor_source = constructor_source
+        self.class_fqdn: str = f'{module_resolver.module.__name__}.{class_name}'
+        self.root_fqdn = root_fqdn
+        self.module_resolver = module_resolver
+        self.class_self_id: str = None
```

Comment:
This suggestions also applies to other places in the code.
```suggestion
self.class_self_id: str
```

Reply:
I cannot change it in other places, it breaks the code. Here it is ok